### PR TITLE
Add Silicon Law cues to Every method a Silicon can have their laws change 

### PR DESCRIPTION
**Pull request by poklj**
## About the PR

All silicons will now get notified if their laws change via any method their laws CAN change, Mostly by adding Cues to ION law changes and Admin Law changes
## Why / Balance

Discussion from #32625 showed that the QoL introduced in that change should also be extended to Borgs getting Ion Laws to ensure a player is aware their laws changed via a sound cue.
## Technical details

Moved all instances of `_roles.MindPlaySound` into its own method (`CueEntityMind`) for easier usage and to keep the relevant code more DRY, this includes For Emag and for the Law board Upload cue introduced in #32625.

Added a call to the new `CueEntityMind` for `OnIonStormLaws` as well as any lawchanges done through `SiliconLawEui`, though this is likely pending a more bespoke Admin Panel change to allow an admin to more directly notify a silicon utilizing any sound they wish (presently defers to the `SiliconLawProviderComponent` already stuck to the target player)
## Media

https://github.com/user-attachments/assets/0a313a5d-dda3-45d6-a805-bf22c9be7242

## Requirements

- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

**Changelog**

:cl: Kickguy223
- add: Ion Storms now notify a Borg that their laws have changed via a sound cue.

### CODE REVIEW

File: Content.Server/Silicons/Laws/SiliconLawEui.cs [hunk 1]
```diff
@@ -52,8 +52,12 @@ public override void HandleMessage(EuiMessageBase msg)
             return;
 
         var player = _entityManager.GetEntity(message.Target);
+        var playerProviderComp = _entityManager.GetComponentOrNull<SiliconLawProviderComponent>(player);
 
         _siliconLawSystem.SetLaws(message.Laws, player);
+
+        if (playerProviderComp != null)
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/SiliconLawEui.cs:59]:
```suggestion
        if (TryComp<SiliconLawProviderComponent>(player, out var playerProviderComp))
```
use TryComp for this

File: Content.Server/Silicons/Laws/SiliconLawEui.cs [hunk 2]
```diff
@@ -52,8 +52,12 @@ public override void HandleMessage(EuiMessageBase msg)
             return;
 
         var player = _entityManager.GetEntity(message.Target);
+        var playerProviderComp = _entityManager.GetComponentOrNull<SiliconLawProviderComponent>(player);
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/SiliconLawEui.cs:55]:
```suggestion
```

**Review by slarticodefast — CHANGES_REQUESTED** [review]
Could you adjust the documentation for the `LawUploadSound` datafield to account for this change? Currently it only mentions AI lawboards.

File: Content.Server/Silicons/Laws/SiliconLawEui.cs [hunk 3]
```diff
@@ -52,8 +52,10 @@ public override void HandleMessage(EuiMessageBase msg)
             return;
 
         var player = _entityManager.GetEntity(message.Target);
-
         _siliconLawSystem.SetLaws(message.Laws, player);
+
+        if (_entityManager.TryGetComponent<SiliconLawProviderComponent>(player, out var playerProviderComp))
+            _siliconLawSystem.CueEntityMind(player, playerProviderComp?.LawUploadSound);
```
Comment (slarticodefast) [inline Content.Server/Silicons/Laws/SiliconLawEui.cs:58]:
```suggestion
            _siliconLawSystem.CueEntityMind(player, playerProviderComp.LawUploadSound);
```
playerProviderComp is never null in the if clause

File: Content.Shared/Silicons/Laws/Components/SiliconLawProviderComponent.cs [hunk 4]
```diff
@@ -24,7 +24,7 @@ public sealed partial class SiliconLawProviderComponent : Component
 
     /// <summary>
     /// The sound that plays for the Silicon player
-    /// when the particular lawboard has been inserted.
+    /// When the Law change is processed for the provider.
```
Comment (slarticodefast) [inline Content.Shared/Silicons/Laws/Components/SiliconLawProviderComponent.cs:27]:
```suggestion
    /// when the law change is processed for the provider.
```

**Review by slarticodefast — APPROVED** [review]
A good addition!
Code looks good and the in-game test went well.
Thank you for your contribution!

**Review by PJB3005 — COMMENTED** [review]
This should probably be part of `NotifyLawsChanged`?

**Comment by slarticodefast** [comment]
> This should probably be part of NotifyLawsChanged?

Oh, yeah. That would be better.

**Comment by poklj** [comment]
> This should probably be part of `NotifyLawsChanged`?

So I've attempted to make this change however i've run into a conundrum. While moving the Cuing of the Entites mind into `NotifyLawsChanged` works for Most instances of how laws are updated, it causes a particular issue with AI Law uploading.

The issue arises due to how the OnUpdaterInsert works, in that the Upload console is given a `SiliconLawUpdaterComponent` with a list of Entities it's in charge of updating which it runs through and calls `SetLaws` with the new LawSet from the `SiliconLawProviderComponent` inserted into it along with each Entities ID's it's managing.

`SetLaws` at that point however has no understanding of the Law Provider that was inserted into the Upload Computers entity, causing it to instead use the Provider on the Target, which is entirely unassociated with the lawset being upload and will no longer have the lawboards specified Cue.

This Can be solved by giving SetLaws an optional parameter to defer to to solve this problem however that would involve a bit more parameter passing as well as modifying the signature of SetLaws

**Comment by slarticodefast** [comment]
Yeah, I had another look at this as well and having a separate helper function is a little cleaner than passing additional optional parameters in my opinion. So I think this solution is good to merge as it is.

**Review by slarticodefast — APPROVED** [review]

**Comment by chromiumboy** [comment]
> > This should probably be part of `NotifyLawsChanged`?
> 
> So I've attempted to make this change however i've run into a conundrum. While moving the Cuing of the Entites mind into `NotifyLawsChanged` works for Most instances of how laws are updated, it causes a particular issue with AI Law uploading.
> 
> The issue arises due to how the OnUpdaterInsert works, in that the Upload console is given a `SiliconLawUpdaterComponent` with a list of Entities it's in charge of updating which it runs through and calls `SetLaws` with the new LawSet from the `SiliconLawProviderComponent` inserted into it along with each Entities ID's it's managing.
> 
> `SetLaws` at that point however has no understanding of the Law Provider that was inserted into the Upload Computers entity, causing it to instead use the Provider on the Target, which is entirely unassociated with the lawset being uploaded and will no longer have the lawboards specified Cue.
> 
> This Can be solved by giving SetLaws an optional parameter to defer to to solve this problem however that would involve a bit more parameter passing as well as modifying the signature of SetLaws

Hmm. OK, I see your point. But I think that overall it would be better to have everything under NotifyLawsChanged like PJB suggested and pass the optional arguments around rather than having to pair it with a separate function for triggering the sound across the codebase

**Review by chromiumboy — APPROVED** [review]
OK, looks good. I'll just get Slarti to check and make sure they agree to the changes

**Comment by slarticodefast** [comment]
Looks good to me. Thank you for your contribution, this is a great QOL addition.

**Pull request merged**